snooty/postprocess.py

```diff
@@ -301,6 +303,66 @@ def populate(self, fileid_stack: FileIdStack, node: n.Node) -> None:
         node.refuri = refuri
 
 
+class ContentsHandler:
+    class HeadingNode(NamedTuple):
+        depth: int
+        id: str
+        title: Sequence[n.InlineNode]
+
+    def __init__(self, diagnostics: Dict[FileId, List[Diagnostic]]) -> None:
+        self.contents_depth = sys.maxsize
+        self.current_depth = 0
+        self.has_contents_directive = False
+        self.headings: List[ContentsHandler.HeadingNode] = []
+        self.diagnostics = diagnostics
+
+    def reset(self, fileid_stack: FileIdStack, page: Page) -> None:
+        self.contents_depth = sys.maxsize
+        self.current_depth = 0
+        self.has_contents_directive = False
+        self.headings = []
+
+    def finalize_headings(self, fileid_stack: FileIdStack, page: Page) -> None:
+        if not self.has_contents_directive:
+            return
+
+        if isinstance(page.ast, n.Root):
+            heading_list = [
+                {
+                    "depth": h.depth,
+                    "id": h.id,
+                    "title": [node.serialize() for node in h.title],
+                }
+                for h in self.headings
+                if h.depth - 1 <= self.contents_depth
+            ]
+            if heading_list:
+                page.ast.options["headings"] = heading_list
+
+    def enter_section(self, fileid_stack: FileIdStack, node: n.Node) -> None:
+        if isinstance(node, n.Section):
+            self.current_depth += 1
+
+    def exit_section(self, fileid_stack: FileIdStack, node: n.Node) -> None:
+        if isinstance(node, n.Section):
+            self.current_depth -= 1
+
+    def __call__(self, fileid_stack: FileIdStack, node: n.Node) -> None:
+        if isinstance(node, n.Directive) and node.name == "contents":
+            self.has_contents_directive = True
+            self.contents_depth = int(node.options.get("depth", sys.maxsize))
```

Comment:
Hmmm. What happens if there's multiple `contents` directives in a page?

One reasonable option might be to skip all but the first

Reply:
This is a great point. Is there any use case for having multiple `contents` directives, or does have to do more with handling edge cases?

Reply:
It seems like there are cases of pages with 2 `contents` directive on the page ([example](https://docs.mongodb.com/manual/meta/aggregation-quick-reference/) - "On this Page" vs. "In this Section")

Reply:
Oy — the next-gen design for "On this page" definitely doesn't support something like this. Given that, I think Andrew's approach to only honor the first instance makes sense.

Reply:
Hmmm, if this is something writers may reasonably expect to work based on sphinx, we should probably raise a warning to let them know.